File: world_simulator_cpp/Tortoise.cpp
#include<iostream>
#include<time.h>
#include "World.h"
#include "Tortoise.h"
using namespace std;

Tortoise::Tortoise(World* currentWorld, int X_coordinate, int Y_coordinate) : Animal(currentWorld, X_coordinate, Y_coordinate, 2, 1, 'T') {

}

void Tortoise::Action() {
	if (this->GetAge() == 0)
		return;
	int move = rand() % 4;
	if (move == 3)
		Move();
}
void Tortoise::Collision(Organism* predator, Organism* prey) {
	if (predator->GetToken() != prey->GetToken()) {
		if (predator->GetStrength() > 5) {
			cout << predator->GetToken() << " attacks " << prey->GetToken() << " on x: " << prey->GetX() << " y: " << prey->GetY() << endl;
			predator->ChangeLocation(prey->GetX(), prey->GetY());
			this->GetWorld()->DelFromWorldMap(prey);
			this->GetWorld()->DelFromOrganismList(prey);
			predator->Draw();
		}
		else
			cout << prey->GetToken() << " defends attack from " << predator->GetToken() << endl;
	}
	else {
		if (predator->GetAge() > 3 && prey->GetAge() > 3)
			this->currentWorld->GiveBirth(predator);
	}
}

File: simulator/Organisms/Animals/Tortoise.py
from simulator.Organisms.Animal import Animal
import random


class Tortoise(Animal):
    def __init__(self, current_world, x_coordinate, y_coordinate):
        super().__init__(current_world, x_coordinate, y_coordinate, 2, 1, True, 'T', 'simulator\\images\\tortoise.png')

    def action(self):
        if self.get_age() != 0:
            if random.randint(0, 3) == 3:
                self.move()

    def collision(self, predator):
        text = ""
        if predator.get_token() != self.get_token():
            text += str(predator.get_token()) + " attacks " + str(self.get_token()) + " "
            if predator.get_strength() > 5:
                text += str(predator.get_token()) + " wins"
                self.get_world().add_log(text)
                tmp_x = self.get_x()
                tmp_y = self.get_y()
                self.get_world().del_organism(self)
                predator.new_location(tmp_x, tmp_y)
            else:
                text += "T defends"
                self.get_world().add_log(text)
        elif self.get_age() > 3 and predator.get_age() > 3:
            self.get_world().new_organism(self)
